src/ucs/datastruct/conn_match.c

```diff
@@ -0,0 +1,211 @@
+    [UCS_CONN_MATCH_QUEUE_EXP]   = "expected",
+    [UCS_CONN_MATCH_QUEUE_UNEXP] = "unexpected"
+};
+
+
+void ucs_conn_match_init(ucs_conn_match_ctx_t *conn_match_ctx,
+                         const ucs_conn_match_ops_t *ops)
+{
+    kh_init_inplace(ucs_conn_match, &conn_match_ctx->hash);
+    conn_match_ctx->ops = *ops;
+}
+
+void ucs_conn_match_cleanup(ucs_conn_match_ctx_t *conn_match_ctx)
+{
+    char dest_addr_str[UCS_CONN_MATCH_ADDR_STR_MAX];
+    ucs_conn_t conn;
+    ucs_conn_match_addr_t conn_match_addr;
+
+    kh_foreach(&conn_match_ctx->hash, conn_match_addr, conn, {
+        if (!ucs_hlist_is_empty(&conn.conn_q[UCS_CONN_MATCH_QUEUE_EXP])) {
+            ucs_diag("match_ctx %p: addr %s expected queue is not empty",
+                     conn_match_ctx,
+                     conn_match_ctx->ops.addr_str(&conn_match_addr, dest_addr_str,
+                                                  UCS_CONN_MATCH_ADDR_STR_MAX));
+        }
+        if (!ucs_hlist_is_empty(&conn.conn_q[UCS_CONN_MATCH_QUEUE_UNEXP])) {
+            ucs_diag("match_ctx %p: addr %s unexpected queue is not empty",
+                     conn_match_ctx,
+                     conn_match_ctx->ops.addr_str(&conn_match_addr, dest_addr_str,
+                                                  UCS_CONN_MATCH_ADDR_STR_MAX));
+        }
+
+        ucs_free(conn_match_addr.addr);
+    })
+    kh_destroy_inplace(ucs_conn_match, &conn_match_ctx->hash);
+}
+
+static ucs_conn_t*
+ucs_conn_match_get_conn(ucs_conn_match_ctx_t *conn_match_ctx,
+                        const ucs_conn_match_addr_t *dest_addr)
+{
+    char dest_addr_str[UCS_CONN_MATCH_ADDR_STR_MAX];
+    ucs_conn_t *conn;
+    ucs_conn_match_addr_t *key;
+    khiter_t iter;
+    int ret;
+
+    iter = kh_put(ucs_conn_match, &conn_match_ctx->hash, *dest_addr, &ret);
+    if (ucs_unlikely(ret == UCS_KH_PUT_FAILED)) {
+        ucs_fatal("match_ctx %p: kh_put failed for %s",
+                  conn_match_ctx,
+                  conn_match_ctx->ops.addr_str(dest_addr, dest_addr_str,
+                                               UCS_CONN_MATCH_ADDR_STR_MAX));
+    }
+
+    conn = &kh_value(&conn_match_ctx->hash, iter);
+
+    if (ret == UCS_KH_PUT_KEY_PRESENT) {
+        return conn;
+    }
+
+    /* initialize match list on first use */
+    conn->next_conn_sn = 0;
+    ucs_hlist_head_init(&conn->conn_q[UCS_CONN_MATCH_QUEUE_EXP]);
+    ucs_hlist_head_init(&conn->conn_q[UCS_CONN_MATCH_QUEUE_UNEXP]);
+
+    key       = &kh_key(&conn_match_ctx->hash, iter);
+    key->addr = ucs_malloc(key->length, "conn match address");
+    if (key->addr == NULL) {
+        ucs_fatal("match_ctx %p: failed to allocate memory for %s address",
+                  conn_match_ctx,
+                  conn_match_ctx->ops.addr_str(dest_addr, dest_addr_str,
+                                               UCS_CONN_MATCH_ADDR_STR_MAX));
+    }
+    ucs_assert(key->length == dest_addr->length);
+    memcpy(key->addr, dest_addr->addr, key->length);
```

Comment:
maybe avoid memcpy, and if user already has address somewhere per connection, just use it?